workflow/controller/operator.go

```diff
@@ -1678,6 +1687,29 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 	return node, nil
 }
 
+// Checks if the template has exceeded its deadline
+func (woc *wfOperationCtx) checkTemplateTimeoutDuration(tmpl *wfv1.Template, node *wfv1.NodeStatus) (*time.Time, error) {
+	if node == nil {
+		return nil, nil
+	}
+
+	if tmpl.TimeoutDuration != nil {
+		tmplMaxDuration, err := parseStringToDuration(tmpl.TimeoutDuration.String())
+		if err != nil {
+			return nil, err
+		}
+
+		deadline := node.StartedAt.Add(tmplMaxDuration)
+
+		if node.Phase == wfv1.NodePending && time.Now().After(deadline) {
+			return nil, fmt.Errorf("%s %s exceeded its deadline", node.Name, node.Type)
+		}
```

Comment:
Why is this only when `node.Phase == wfv1.NodePending`? Shouldn't this check be made regardless of phase?

Reply:
This code is handling mainly for nodes that are pending because of waiting for sync lock or parallelism. Rest of state will be handle in createPod function